discovery/db_discovery.go

```diff
@@ -19,38 +18,37 @@ import (
 	"github.com/golang/glog"
 )
 
-var cacheRefreshInterval = 1 * time.Hour
-var getTicker = func() *time.Ticker {
-	return time.NewTicker(cacheRefreshInterval)
+type ticketParamsValidator interface {
+	ValidateTicketParams(ticketParams *pm.TicketParams) error
+}
+
+type roundsManager interface {
+	LastInitializedRound() *big.Int
 }
 
 type DBOrchestratorPoolCache struct {
-	node *core.LivepeerNode
+	store                 common.OrchestratorStore
+	lpEth                 eth.LivepeerEthClient
+	ticketParamsValidator ticketParamsValidator
+	rm                    roundsManager
 }
 
-func NewDBOrchestratorPoolCache(node *core.LivepeerNode) *DBOrchestratorPoolCache {
-	if node.Eth == nil {
-		glog.Error("Could not refresh DB list of orchestrators: LivepeerNode nil")
+func NewDBOrchestratorPoolCache(store common.OrchestratorStore, lpEth eth.LivepeerEthClient, rm roundsManager, ticketParamsValidator ticketParamsValidator) *DBOrchestratorPoolCache {
+	if lpEth == nil {
+		glog.Error("could not create DBOrchestratorPoolCache: LivepeerEthClient is nil")
 		return nil
 	}
 
-	_ = cacheRegisteredTranscoders(node)
-
-	ticker := getTicker()
-	go func(node *core.LivepeerNode) {
-		for _ = range ticker.C {
-			err := cacheRegisteredTranscoders(node)
```

Comment:
While we no longer need to query for on-chain pool at regular intervals, we still want to call `GetOrchestrator()` on known orchestrators at regular intervals to fetch off-chain information. We no longer do that in the current set of commits since this code was removed.

Reply:
I feel like there's several options here

1. Start a simple go routine that queries the DB for all orchestrators that are active (we don't care about inactive ones) , fetch their off-chain info and write the price to the DB. This routine runs seperately from any other functionality (similar to the old caching loop). 

2. Remove MaxPrice filter when querying the DB, solely query for active orchestrators and only worry about filtering for max price in the `predicate` function. This would gives us up-to-date pricing information for active O's on each `OrchestratorPool.GetOrchestrators()`

3. Work out a `OrchestratorInfoPoller` implementation as described in [this doc](https://hackmd.io/SiCcxJXPTTeSJ1HsZEekmA)

We could go for 1 or 2 and implement 3 at a later stage, or bite the bullet and go for 3 now. But since you want to get the release out and `OrchestratorInfoPoller` needs some extra thought I would suggest going for option 2.  WDYT?

Reply:
I would go with option 1. We ultimately want the orchestrator to fetch off-chain prices at a regular interval so that it can filter by price while constructing its working set for a stream so that it only needs to send requests to orchestrators that meet its price requirements. We also fetch & validate the orchestrator's info in `GetOrchestrators()`, but this is just a sanity check in case the orchestrator's price is no longer valid. Option 1 preserves this behavior and then at a later point we can consider option 3 as a code quality improvement.

Reply:
Added in 0be3692

Not exactly the same as before because we need a `common.Broadcaster` instance to pass into `server.GetOrchestratorInfo` to generate the request, but quite similar.

Reply:
@kyriediculous Just read the writeup for the OrchestratorInfoPoller - good stuff, thanks for sharing! Just one point I wanted to ask about:

> Benefits of caching OrchestratorInfo
> Reduces the amount of requests sent over the wire by caching, rather than fetching the data each time selectOrchestrator is ran

Does this mean the reducing number external calls to reconstruct the state of the current orchestrator set - the current intent of the orch DB polling?

Or is the proposal here more around actually avoiding *all* OrchestratorInfo requests at selection time?

Calling OrchestratorInfo at selection time is still something we should still do; as noted in the hackmd, the data can easily go stale, orchestrators may become unavailable, etc.

While the poller's responsibility is to update the orch DB to help filter out the list as a first pass (minimizes blockchain lookups, excludes clearly unfit orchestrators), we should still be calling OrchestratorInfo at selection time to ensure up-to-date information on pricing and availability.

Reply:
@j0sh My idea initially was to abolish any network requests when selecting orchestrators but when I took a deeper look it's not as easy. In order to not have stale data we would have to (apart from polling at regular intervals) write any `OrchestratorInfo` response included in a call to `SubmitSegment` to the database, injecting the DB as a dependency in `server` doesn't seem quite feasible however. So far the only option to achieve that seems to be adding a method to `core.Broadcaster`, then in submit segment we can call `BroadcastSession.BroadCaster.UpdateOrchestratorInfo`  which under the hood just calls `db.UpdateOrch`.

Alternatively, we can add an `OrchestratorStore` instance to every `BroadcastSession`, which is easy to do because `broadcaster.selectOrchestrator` takes in `LivepeerNode` as an argument on which we can access `db`. 

This structure would prevent stale data due to polling and a feedback loop I think.

Reply:
> My idea initially was to abolish any network requests when selecting orchestrators

Orchestrator selection is usually not time sensitive [1], so I'm not really sure this needs to be a goal? What else are we trying to achieve with this change?

The current approach (before this PR) is a good compromise between the trade-offs of immediacy and freshness. Periodic polling fetches off-chain information, which helps filter out orchestrators as a first pass during selection time. Then actual selection is an opportunity to confirm pricing and availability as needed, without running the risk of having stale data.

[1] The only time O selection blocks a pending segment is for the very first request establishing a session with the HTTP ingest - but there are ways to work around this. RTMP and mid-session refreshes don't typically have issues with refreshes or O selection, as long as the working set of Os is sufficiently large (>2).

Reply:
> The only time O selection blocks a pending segment is for the very first request establishing a session with the HTTP ingest 

Ah thanks , I did not realise this. In that case the current approach is as you say a good compromise indeed.